packages/backend/src/graphql/mutations/tiles/upsert-table-collaborator.ts

```diff
@@ -0,0 +1,85 @@
+import { ITableCollabRole } from '@plumber/types'
+
+import { BadUserInputError } from '@/errors/graphql-errors'
+import { getOrCreateUser } from '@/helpers/auth'
+import { validateAndParseEmail } from '@/helpers/email-validator'
+import logger from '@/helpers/logger'
+import TableCollaborator from '@/models/table-collaborators'
+
+import type { MutationResolvers } from '../../__generated__/types.generated'
+
+const upsertTableCollaborator: MutationResolvers['upsertTableCollaborator'] =
+  async (_parent, params, context) => {
+    const { tableId, email, role } = params.input as {
+      tableId: string
+      email: string
+      role: ITableCollabRole
+    }
+
+    const validatedEmail = await validateAndParseEmail(email)
+    if (!validatedEmail) {
+      throw new BadUserInputError('Invalid collaborator email')
+    }
+
+    if (context.currentUser.email === validatedEmail) {
+      throw new BadUserInputError('Cannot change own role')
+    }
+
+    if (role === 'owner') {
+      throw new BadUserInputError('Cannot set collaborator role as owner')
+    }
+
+    await TableCollaborator.hasAccess(context.currentUser.id, tableId, 'editor')
```

Comment:
nit: should this be checked first similar to `deleteTableCollaborator`?

Reply:
checking bad inputs that dont need db queries first before checking permissions so the short circuiting will be faster